oeps/best-practices/oep-0051-bp-conventional-commits.rst

```diff
@@ -148,6 +150,8 @@ Choosing the commit type label to use for a commit is important, but it's only t
 
 **feat vs fix**: some user-visible changes to features could be classified as "feat" or "fix".  Choose "feat" if the change adds to the set of features.  Choose "fix" if the change affects how a feature behaves.  Yes, this is still subjective.
 
+**squash vs temp vs unknown or missing prefix**: commits that are not intended to be merged could use no prefix or an unknown prefix to communicate to a linter that a PR is not ready to merge. This is nice, but it also forces reviewers to more closely track why a PR is failing throughout the review process. An alternative is to categorize a commit that is *not* intended to be squashed before merge by using "squash". Do *not* use "temp" or "fix" for this use case, since these categories convey a different meaning once merged, and are intended to be merged. Note that if you use "squash" commits, rather than omitting the prefix, you will no longer get reminded to squash through a failed PR.
```

Comment:
The purpose of a pre-merge check is to catch things that should not be merged to the main branch; this is true of code linting, unit tests, and now commit linting.  Making commitlint "turn a blind eye" to commits that _expressly_ should not be merged defeats its purpose.  It's much better for a reviewer to keep track of why a PR is failing checks than to track whether is is passing checks that it shouldn't; pre-merge checks should _only_ pass when the PR is, in fact, ready to merge.

Put it like this:
- You have a test case you want to be sure to cover before you merge so you're going to commit a stub placeholder version now and fill it in before you merge; is it better to make that stub pass so the unit test check stays green and it's easier to see if other tests break or should you make it fail so the unit test check ensures that placeholder will not get merged?  I assert that the latter is much better practice.
- You have a commit that you want to squash before you merge; is it better to use a commit type that will pass the check so your PR stays green or to use _anything else_ so the check ensures that commits you don't intend to merge don't end up on the main branch?  Again, I assert the latter is much better practice.

I think a `squash:` type is a great suggestion and we should use it.  However, as its express purpose is to mark commits that are not intended to be merged, commitlint _must_ flag such commits to reviewers since the PR, as it stands, is not quite ready to merge; the reviewer can, of course, respond "approved assuming the `squash:` commits are squashed.", just as we sometimes do when a PR needs minor changes, is failing a unit test, or any other pre-merge check.

Reply:
[Note: The following was written assuming commitlint becomes required at some point.]

**Option 1: commitlint fails for `squash:` commits, "Squash and merge" is _disabled_:**

Pros:

- `squash:` commits can't be merged.
- Master can't be broken with a bad message from "Squash and merge".

Cons:

- Authors must squash locally to get a green build.
  - This adds another round of running tests, which is slow on many cases.
  - Re-running tests puts pressure on authors to squash early, but squashing early wastes reviewer time if they want to see diffs for "minor" final changes.
  - Github's popular "Squash and merge" option would need to be disabled for commitlint to become mandatory. 
  - Minor annoyance to both author and reviewers (in my opinion) that any PR that requires squashing must be red throughout the review process because squashing should be the final step. 
- Even if we add `squash:` to the OEP, there would be no linting to ensure consistency on this particular option.

**Option 2: commitlint fails for `squash:` commits, "Squash and merge" remains _enabled_:**

Pros:

- Developers can continue to use Github's "Squash and merge".
  - Tests don't need to be re-run just to squash. 
  - Authors are not incentivized to squash early, enabling a more efficient reviewer experience.
  - Reviewers and authors can get to a green build earlier in the process that covers most cases. 
- A green commitlint signifies all commits conform to the OEP and provide meaning.

Cons:

- It remains possible that `squash:` commits could be merged.
  - Although not ideal, the commit prefix probably still provides enough meaning when this does happen.
- Master could be broken with a bad message using "Squash and merge"
  - Again, although not ideal, this is similar to not forcing every PR to be up-to-date with master before merging. In the case of not requiring a rebase off master, we trade-off developer efficiency for the common case against ensuring that master can't be broken with a PR that would have failed unit tests had it been re-based ahead of time.

I prefer Option 2, but there are pros and cons to both options.

Reply:
When or whether we start requiring commitlint for merging and whether we would then continue to allow squash-and-merge if we do are separate from whether commitlint should flag commits explicitly marked as do-not-merge or should pass in spite of them.

Do we really consider `squash:` commits to be "good enough to merge"?

As a reviewer, I want all outstanding issues that can be automatically caught to be spelled out by the pre-merge checks.  Often, it is fine to review a PR before every issue has been resolved, but if there's lint issue, or a unit test that needs to be finished or updated before merging, I want those to be flagged; if there are commits whose only meaning is "this should be squashed away", I want those to be flagged too.  It's fine to say "approved pending...", and in this case, it's fine for the author to resolve these commits with squash-and-merge.

As a reviewer, I want all-checks-passing to be as strong a statement about the PR's readiness to be merged as possible, I would _much_ rather parse failing checks than have to second-guess the actual readiness of a passing check.

**THE** big "con" for option 2 as outlined above is that all-checks-passing would not indicate whether a PR is actually ready to merge (unless we decide that `squash:` commits _are_ good enough to merge); "getting to a green build sooner" is not a "pro" if the PR still has outstanding issues.

But, again, we do not currently require commitlint to pass for merging, and changing that is a much bigger discussion with much wider implications.  I am advocating for...

**Option 3: commitlint fails for `squash:` commits, "Squash and merge" remains enabled for the time being:**

Pros:
- Developers are made aware that they shouldn't use other types for commits that are not intended to merge to the main branch.
- Commits that are not truly meaningful are much less likely to be merged to the main branch.
- Developers can continue to use Github's "Squash and merge".
- Tests don't need to be re-run just to squash.
- Authors can choose whether to interactively rebase their changes into commits that capture complete changes or leave them to squash later if that would be easier to review.
- Reviewers and authors can see exactly what issues should be resolved before merging.
- All-checks-passing makes a strong statement about the PR's readiness to merge.

Cons:
- Reviewers may be more frequently reviewing PRs before all checks pass.
- It remains possible that `squash:` commits could be merged.
    - Although not ideal, the commit prefix probably still provides enough meaning when this does happen.
- Master could end up with a bad message edited using "Squash and merge".
    - Again, although not ideal, this is similar to not forcing every PR to be up-to-date with master before merging. In the case of not requiring a rebase off master, we trade-off developer efficiency for the common case against ensuring that master can't be broken with a PR that would have failed unit tests had it been re-based ahead of time.

Reply:
@inventhouse: Agreed. I was also about to write a comment stating that I'd update this PR to remove language around how linting will be performed, and move that discussions elsewhere.

This debate can be taken up on the linter before or after it becomes required. Also, it would be nice to have a global decision, but the decision could be different for different repos.

Reply:
I think this is an important discussion, but I agree that this OEP is about the format and types, and specifically not about the tooling, marking this as resolved.

Reply:
@inventhouse: This thread is no longer marked as resolved so it is easier for others to see and contribute to. Others clearly have thoughts on this topic, even though we are hoping to merge this PR without having to make a final decision on this separate topic (i.e. whether to pass or fail linting on "squash" commits).

Reply:
Might it be possible to have commitlint _warn_ (but not fail) for `squash`?  (This would, of course, still leave unanswered the question of “what does a required commitlint check even mean if the message can be completely rewritten after the check)

Reply:
1. Warning sounds more appropriate than error.
2. The required check still ensures that the author knows about and is generally using conventional commits. If it is rewritten afterwards, it is likely to use an existing comment that already conformed. Could some intentionally cause a problem? Sure, but this seems like a very low risk.
3. It is important to note that the pass/fail decision is easily reversible. The only issue is managing the change in expectations.

Reply:
2) “could someone _make a mistake_” is the more likely issue, and, again, how is a non-binding “required” check actually different than than the current state?

Reply:
The goal is to get people to write  conventional commits. Ned provides data about the growth of their usage, but we don’t have full compliance. I bet making it a non-binding “required” check would spike this data and give us what we are looking for. Also, because it would minimally disturb the typical developer workflow, we could more quickly make it required without upsetting developers, and more quickly get compliance on using conventional commits.

@nedbat: I’d love to hear where you stand in this given your goal to get people using conventional commits.

Reply:
Jumping in late, but it sounds like this PR is tackling the problem of devs taking the path of least resistance[^1] by using inappropriate commit prefixes.

The proposed solution is to (a) Create a "blessed illegal" commit prefix, as Tim well put it, and (b) exercise one of Options 1–3 to make commitlint and our GitHub merge strategies create a better path of least resistance.

Very controversial opinion[^2]:

**Option 4**: Make `temp` harder to use.

This can be done by requiring the word "until" in the commit message and having very good commitlint console output for the dev.

As a dev, I prefer being taught the right way than having to keep track of one more exception in my head.[^3]

For `fix`: Odds are, it's probably being used correctly; it just also happens to be not intended for merging.

And that way, with Option 4, we don't need the `squash` merge prefix.

[^1]: This is a good thing! Developers should not be spending thinking time taking paths of non-least resistance. Plus, it provides architects the challenge of crafting the path of least resistance so the commit prefixes are useful.

[^2]: This is also a _very_ weak opinion. I'd be happy to go either way as a dev.

[^3]: Fair, having to use "until" only for `temp` is an exception as well, but it is an exception which is crafted to help the dev adhere to a larger principle, rather than create an exception to the principle.

Reply:
@pshiu `squash` has a different meaning than `temp`:

> **squash vs temp or fix**: use "squash" when the *commit* is temporary, and you intend to squash the commit into a larger commit with a different semantic label. Use "temp" when the *change* is temporary, but you intend to merge the commit. Again, use "squash" over "fix" when the *commit* is temporary and is not intended to be merged.

Though I like the idea of expiration dates on `temp`, that'd be a different proposal for a different problem

Reply:
Looks like it would be pretty straightforward to create a custom rule to warn for `squash` commits: [Commitlint Local Plugins](https://commitlint.js.org/#/reference-plugins?id=local-plugins)

Reply:
@inventhouse: I'm not sure if we'd need a custom rule. Wondering if something like this would work: https://github.com/edx/edx-lint/pull/215/files

Reply:
@robrap I don't expect you're allowed to have two rule entries with the same key; I think it would have to be a different rule name.

Reply:
@inventhouse: I'm guessing you are right, but it also shows the pattern for how we could use the custom rule. I wasn't clear on that at first.